src/main/java/org/springframework/data/couchbase/repository/query/AbstractN1qlBasedQuery.java

```diff
@@ -71,14 +78,22 @@ protected AbstractN1qlBasedQuery(CouchbaseQueryMethod queryMethod, CouchbaseOper
    */
   protected abstract boolean useGeneratedCountQuery();
 
-  protected abstract Statement getStatement(ParameterAccessor accessor, Object[] runtimeParameters);
+  protected abstract Statement getStatement(ParameterAccessor accessor, Object[] runtimeParameters, ReturnedType returnedType);
 
   protected abstract JsonValue getPlaceholderValues(ParameterAccessor accessor);
 
   @Override
   public Object execute(Object[] parameters) {
     ParametersParameterAccessor accessor = new ParametersParameterAccessor(queryMethod.getParameters(), parameters);
-    Statement statement = getStatement(accessor, parameters);
+
```

Comment:
I tried moving all of the mapping to DTO to couchbaseOperations but due interface mapping, it still happens on AbstractN1QLBasedQuery because I couldn't pass down the domain type and interface return type until findByN1QL method without breaking api, we'll discuss more on this tomorrow with Simon